algorithmic-strategies/dynamic-programming.md

```diff
@@ -0,0 +1,302 @@
+
+And how do we find the maximum subsequence of the remaining strings? Why with recursion!
+
+![The call tree for finding the longest common subsequence for the strings "ace" and "ae". Because there are three branches at each level, it spreads very quickly. There are large repeated sections of the call tree.](../assets/algorithmic-strategies_dynamic-programming_lcs.png)  
+_Fig. If the recursive explosion of Fibonacci seemed bad, get a load of this! Crossed-out nodes indicate calls that result in a 0 length due to at least one of the input strings being empty. The numbers indicate the maximum value being returned back from a particular branch of execution. Repeated regions are outlined in yellow._
+
+The explosion of calls in this diagram puts Fibonacci to shame! But for small examples, like in the example table above, it can be manageable with a traditional recursive implementation.
+
+```python
+def lcs(str1, str2):
+    if not str1 or not str2:
+        return 0
+
+    # split the first character from the rest
+    first1 = str1[0]
+    rest1 = str1[1:]
+    first2 = str2[0]
+    rest2 = str2[1:]
+
+    # is this spot a match?
+    if first1 == first2:
+        current_score = 1
+    else:
+        current_score = 0
+
+    # the result for this position is the max of the current score
+    # and each of the maxes from the three possibilities:
+    # 1. advance both characters (adding to score if a match)
+    # 2. advance only the first character
+    # 3. advance only the second character
+    result = max(
+        current_score + lcs(rest1, rest2),
+        lcs(rest1, str2),
+        lcs(str1, rest2)
+    )
+
+    return result
+```
+
+But increasing the input size can slow things down quickly! Consider calculating the longest common subsequence for `"tagacgttagtc"` and `"qaqaqgqtqgqc"`. The traditional recursive solution will take a noticeable amount of time.
+
+From the example call diagram, if we notice that there are many repeated sub-trees (especially the calls to `lcs("ce", "e")`, outlined in yellow), we might think about applying dynamic programming practices to reuse these calculations.
+
+To add memoization to the Fibonacci algorithm, we only had to check whether a single value had been calculated before. But for the longest common subsequence we need to look up a result based on two inputs.
+
+<br />
+
+<details>
+
+<summary>Think about how we might accomplish this, and then click here to review a possible solution. Comments are included to indicate the newly added code.</summary>
```

Comment:
For situations like this, I randomly prefer "Consider" over "Think about."

For me, "thinking about" feels more like musing, like it should be casual, while "considering" has that _weight_, but in a casual way.

Split the first sentence to give the "Consider" step a nice pause.

I won't fight for this suggestion :)

```suggestion
<summary>Consider how we might accomplish this. Then click here to review a possible solution. Comments are included to indicate the newly added code.</summary>
```